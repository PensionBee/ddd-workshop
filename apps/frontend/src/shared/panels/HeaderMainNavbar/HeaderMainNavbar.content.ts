import type { NavMenuItem } from "@shared/components/Nav/Nav.types";

export const NAV_MENU_CONTENT = [
  {
    children: "Our service",
    subList: [
      { children: "How it works", href: "/how-it-works" },
      { children: "Combine", href: "/combine" },
      { children: "Contribute", href: "/contribute" },
      {
        children: "Withdraw",
        subList: [
          {
            children: "About pension withdrawals",
            href: "/pension-withdrawal",
          },
          { children: "Drawdown", href: "/drawdown" },
          { children: "Annuities", href: "/annuity" },
        ],
      },
      {
        children: "Plans",
        subList: [
          { children: "Plans", href: "/plans" },
          {
            children: "Fossil fuel free investing",
            href: "/fossil-fuel-free-pension",
          },
        ],
      },
      { children: "Self-employed", href: "/self-employed-pension" },
      { children: "Fees", href: "/fees" },
    ],
  },
  {
    children: "About us",
    subList: [
      { children: "Vision", href: "/our-vision" },
      { children: "Values", href: "/values" },
      { children: "ESG", href: "/esg" },
      { children: "History", href: "/our-history" },
      { children: "Reviews", href: "/reviews" },
      { children: "Press", href: "/press" },
      { children: "Careers", href: "/careers" },
    ],
  },
  {
    children: "Knowledge",
    subList: [
      { children: "Pensions explained", href: "/pensions-explained" },
      { children: "Pension Academy videos", href: "/pension-academy-videos" },
      { children: "Blog", href: "/blog" },
      { children: "Pension calculator", href: "/pension-calculator" },
      { children: "Drawdown calculator", href: "/drawdown-calculator" },
      { children: "Pension landscape", href: "/pension-landscape" },
    ],
  },
  {
    children: "Help & FAQs",
    subList: [
      { children: "Contact us", href: "/contact-us" },
      { children: "FAQs", href: "/faq" },
      { children: "Terms", href: "/terms" },
      { children: "Privacy policy", href: "/privacy-policy" },
    ],
  },
] satisfies NavMenuItem[];
