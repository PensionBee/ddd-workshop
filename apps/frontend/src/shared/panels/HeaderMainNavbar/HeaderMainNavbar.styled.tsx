import { mediaFrom } from "@shared/styling/mediaQueries";
import styled from "styled-components";

export const StyledMenuDrawer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  background-color: #fff;
  transform: translateX(320px);
  position: fixed;
  top: 0;
  right: 0;
  width: 320px;
  overflow: hidden;
  padding: 50px 0 20px;
  box-sizing: border-box;
  box-shadow: 0 0 0px rgb(174 190 210 / 40%);
  bottom: 0;
  z-index: 1;
  transition: transform 350ms ease, box-shadow 350ms ease;
  max-width: 100%;
  [data-close] {
    position: absolute;
    top: 20px;
    right: 20px;
    display: flex;
  }
  &[data-open="true"] {
    transform: translate(0);
    box-shadow: 0 0 20px rgb(174 190 210 / 40%);
  }
  ${mediaFrom.md} {
    padding: 108px 0 20px;
  }
`;

export const StyledMenuButton = styled.button`
  cursor: pointer;
`;
