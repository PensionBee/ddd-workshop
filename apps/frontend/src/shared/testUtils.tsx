import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
