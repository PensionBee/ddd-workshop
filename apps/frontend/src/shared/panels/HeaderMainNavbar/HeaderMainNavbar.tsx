import { useState } from "react";
import { StyledMenuButton, StyledMenuDrawer } from "./HeaderMainNavbar.styled";
import { LogoSvg, LargeSystemIcon } from "@shared/components/Svg";
import { Button } from "@shared/components/Button";
import { GridContainer } from "@shared/components/Grid";
import { MainNav } from "@shared/components/Nav/MainNav/MainNav";
import { MobileMainMenu } from "@shared/components/Nav/MobileMenu/MobileMenu";
import { Link } from "@shared/components/Link";
import { Box } from "@shared/components/Box";
import { NAV_MENU_CONTENT } from "./HeaderMainNavbar.content";
import { useWindowScroll } from "@shared/hooks/useWindowScroll/useWindowScroll";
import {
  useMediaQueryMd,
  useMediaQueryLg,
} from "@shared/hooks/useMediaQuery/useMediaQuery";

export const HeaderMainNavbar: React.FC = (props) => {
  const windowScroll = useWindowScroll();

  const hasScrolled0 = windowScroll >= 0;
  const hasScrolled100 = windowScroll >= 100;
  const hasScrolled200 = windowScroll >= 200;

  return (
    <>
      <HeaderMainNavbarInner
        data-rc="HeaderMainNavbar"
        isVisible={
          hasScrolled200 || !(hasScrolled0 && hasScrolled100) || !hasScrolled100
        }
        isFixed={hasScrolled200}
        {...props}
      />
      <Box mt={{ xs: 45, lg: 82 }} />
    </>
  );
};

type HeaderMainNavbarProps = {
  isFixed?: boolean;
  isVisible?: boolean;
};

const HeaderMainNavbarInner: React.FC<HeaderMainNavbarProps> = ({
  isFixed = false,
  isVisible = true,
  ...props
}) => {
  const isTablet = useMediaQueryMd();
  const isDesktop = useMediaQueryLg();
  const [menuOpen, setMenuOpen] = useState(false);

  return (
    <Box
      position={isFixed ? "fixed" : "absolute"}
      backgroundColor="white"
      z={2}
      w="100vw"
      top={0}
      boxShadow={isFixed ? "0px 0px 20px rgb(197 225 240 / 40%)" : undefined}
      transition={isVisible ? "opacity 300ms" : undefined}
      opacity={isVisible ? 1 : 0}
      {...props}
    >
      <GridContainer>
        <Box
          d="flex"
          justify="space-between"
          align="center"
          py={{ xs: 5, lg: 15 }}
          w="100%"
        >
          <Box d="flex">
            <Link href="/">
              <LogoSvg
                icon="PensionBeeLogo"
                h={isTablet ? 52 : 37}
                w={isTablet ? 172 : 122}
              />
            </Link>
          </Box>
          {isDesktop ? (
            // Desktop Menu Items
            <>
              <Box ml="auto" mr={{ xs: 20, md: 40 }}>
                <MainNav list={NAV_MENU_CONTENT} />
              </Box>
              <Box d="flex" align="center" colGap={20}>
                <Link href="/logout" color="black" noUnderline>
                  Log Out
                </Link>
                <Button href="/dashboard" w="unset">
                  BeeHive
                </Button>
              </Box>
            </>
          ) : (
            // Mobile Menu Items
            <Box d="flex">
              <StyledMenuButton onClick={() => setMenuOpen(true)}>
                <LargeSystemIcon icon="Menu" />
              </StyledMenuButton>
            </Box>
          )}
        </Box>
      </GridContainer>
      {!isDesktop && (
        // Mobile Menu Drawer
        <Box z={3}>
          <StyledMenuDrawer data-open={menuOpen}>
            <StyledMenuButton onClick={() => setMenuOpen(false)} data-close>
              <LargeSystemIcon icon="Close" w={24} h={24} />
            </StyledMenuButton>
            <MobileMainMenu list={NAV_MENU_CONTENT} />
            <Box p={20} d="flex" rowGap={20} colGap={20}>
              <Box d="flex" flex="1">
                <Button href="/logout" variant="secondary" w="100%">
                  Log Out
                </Button>
              </Box>
              <Box d="flex" flex="1">
                <Button href="/dashboard" w="100%">
                  BeeHive
                </Button>
              </Box>
            </Box>
          </StyledMenuDrawer>
        </Box>
      )}
    </Box>
  );
};
