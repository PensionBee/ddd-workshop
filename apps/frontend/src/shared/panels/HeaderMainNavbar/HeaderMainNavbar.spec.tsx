import { mockMediaMatches, render } from "@shared/testUtils";
import { HeaderMainNavbar } from "./HeaderMainNavbar";

describe("HeaderMainNavbar", () => {
  const TEST_ID = "HeaderMainNavbar";
  it("should render", () => {
    mockMediaMatches("xs");
    const { getByTestId, asFragment } = render(
      <HeaderMainNavbar data-testid={TEST_ID} />
    );
    expect(getByTestId(TEST_ID)).toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  });
});
