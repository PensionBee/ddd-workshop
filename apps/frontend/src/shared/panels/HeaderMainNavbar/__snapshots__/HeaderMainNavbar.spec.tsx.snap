// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderMainNavbar > should render 1`] = `[Function]`;

exports[`HeaderMainNavbar should render 1`] = `[Function]`;
